experimenter/experimenter.py

```diff
@@ -181,45 +176,44 @@ def _generate_standard_pipeline(self, preprocessor: PrimitiveBase, classifier: P
                     preprocessor_used = True
 
-                step_4.add_argument(name=arg, argument_type=ArgumentType.CONTAINER,
+                preprocessor_step.add_argument(name=arg, argument_type=ArgumentType.CONTAINER,
                                     data_reference=data_ref)
-            step_4.add_hyperparameter(name='use_semantic_types', argument_type=ArgumentType.VALUE, data=True)
-            step_4.add_hyperparameter(name='return_result', argument_type=ArgumentType.VALUE, data="replace")
-            step_4.add_output('produce')
-            pipeline_description.add_step(step_4)
+            preprocessor_step.add_hyperparameter(name='use_semantic_types', argument_type=ArgumentType.VALUE, data=True)
+            preprocessor_step.add_hyperparameter(name='return_result', argument_type=ArgumentType.VALUE, data="replace")
+            preprocessor_step.add_output('produce')
+            pipeline_description.add_step(preprocessor_step)
             step_counter += 1
 
-        # Step 5: Classifier
-        step_5 = PrimitiveStep(primitive_description=classifier.metadata.query())
-        for index, arg in enumerate(self._get_required_args(classifier)):
+        # Classifier Step
+        classifier_step = PrimitiveStep(primitive_description=classifier.metadata.query())
+        for arg in self._get_required_args(classifier):
             if arg == "outputs":
-                data_ref = 'steps.3.produce'
+                data_ref = f'steps.{target_step_i}.produce'
             else:
                 # if we haven't used a preprocessor, we have to go back two to get the full dataframe
-                bias = 0 if preprocessor_used else -1
-                data_ref = 'steps.{}.produce'.format(step_counter - 1 + bias)
-            step_5.add_argument(name=arg, argument_type=ArgumentType.CONTAINER,
+                data_ref = f'steps.{step_counter - 1 if preprocessor_used else attributes_step_i}.produce'
+            classifier_step.add_argument(name=arg, argument_type=ArgumentType.CONTAINER,
                                 data_reference=data_ref)
-        step_5.add_hyperparameter(name='use_semantic_types', argument_type=ArgumentType.VALUE, data=True)
-        step_5.add_hyperparameter(name='return_result', argument_type=ArgumentType.VALUE, data="replace")
-        step_5.add_output('produce')
-        pipeline_description.add_step(step_5)
+        classifier_step.add_hyperparameter(name='use_semantic_types', argument_type=ArgumentType.VALUE, data=True)
+        classifier_step.add_hyperparameter(name='return_result', argument_type=ArgumentType.VALUE, data="replace")
+        classifier_step.add_output('produce')
+        pipeline_description.add_step(classifier_step)
         step_counter += 1
 
-        step_counter = self._add_predictions_constructor(pipeline_description, step_counter)
+        step_counter = self._add_predictions_constructor(pipeline_description, step_counter, attributes_step_i)
```

Comment:
Do you know how the add predictions primitive works @orionw? Should I have maybe passed the step index of the targets as the reference to that primitive, rather than the step index of the attributes?

Reply:
I figured it out. I looked at one of JPL's sklearn pipeline examples. They just use the output from step 0 as a reference, and it works.